test/kafka-exactly-once/mzworkflows.py

```diff
@@ -0,0 +1,55 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from typing import List
+
+from materialize.mzcompose import (
+    Kafka,
+    Materialized,
+    SchemaRegistry,
+    Testdrive,
+    Workflow,
+    WorkflowArgumentParser,
+    Zookeeper,
+)
+
+services = [
+    Zookeeper(),
+    Kafka(),
+    SchemaRegistry(),
+    Materialized(),
+    Testdrive(),
+]
+
+
+def workflow_kafka_exactly_once(w: Workflow, args: List[str]):
+    parser = WorkflowArgumentParser(w)
+    parser.add_argument(
+        "--seed",
+        help="an alternate seed to use to avoid clashing with existing topics",
+        type=int,
+        default=1,
+    )
+    args = parser.parse_args(args)
+
+    w.start_services(services=["kafka", "schema-registry", "materialized"])
```

Comment:
I do not think this is safe as in rare cases some of the Java services will fail to start if the previous ones in the sequence have not started already.

We have `w.start_and_wait_for_tcp` that takes a list of actual service objects and starts them **in sequence**, pinging each other as it goes. This is safer than starting the Java parts with start_services as start_services does not guarantee that the previous service has started already before the next one is started.

Reply:
Ooh, thanks!